### ISSUE TITLE
Webdataset (Liaon115M) + Torchlightning (pl.DataModule) with visualizing progressbar during training

### ISSUE THREAD

**Issue by HuangChiEn** [webdataset/webdataset#346]
This issue aims to **end** with the discussion that how to run the webdataset with torchlightning. 

Several concepts should be clarified here : 
1. If we use webdataset, it means the dataset is large-scale (if not, then why bother ?), so the multi-gpu support is definitely needed!! 
2. Yes, dataset length is impotent. No one! yes, **no one want to train on a dataset without knowing the step per epoch in large-scale dataset!!**. So, dataset length (num of data sample in dataset) is definitely needed!! 
3. It's weird that torch IterableDataset doesn't support too much in multi-gpu manner. Because the reason we use IterableDataset just exactly that we're working with the large-scale dataset and Multi-gpu is necessary for large-scale training.

ok, if dataset length is impotent, then how can I get it ?
> while, we wrote a parallel script for scanning the large-scale dataset, and record the image for each tar-file and store in a json file with dict manner : `{'/data/laion115m/00000.tar':1147, '/data/laion115m/00001.tar':..., /data/laion115m/11164.tar' : 551}`.  

---

After some murmur, let's begin 👍 
1. First step, use Webdataset with **toy usage only** (for-loop with single stream iteration)
```
# feed dummy pipe `wds.split_by_worker`, which is the default of workersplit, thus it'll not affect the results
dataset = wds.WebDataset(url, nodesplitter=wds.split_by_worker).shuffle(1000).decode('pilrgb', handler=wds.warn_and_continue).to_tuple("jpg", "txt", handler=wds.warn_and_continue).map(trfs)

# toy usage : 
for batch_id, sample in enumerate(dataset):
      print('ok, good! we can get the sample')
```

2. Wrap webdataset in torch IterableDataset (aims to do nodesplit)
```
from torch.distributed import get_rank, get_world_size

class Iter_ds(torch.utils.data.IterableDataset):
    def __init__(self, urls, trfs, n_sample):
        self.urls = urls
        self.trfs = trfs
        self.n_sample = n_sample
        
    def __len__(self):  
        # let's say i have 100 image totally, 2 gpus, batch_size = 4.
        # then n_step per epoch should be : `100 // (2 * 4) = 17`, and last batch doesn't fill with 4 samples.
        # in here, we directly control how many n_step by __len__. 
        # for the above example, in here should be `100 // 2`, then torch dataloader will try to divided this number by batch_size automatically ~ (since we setup batch_size in torch dataloader)  
        return self.n_sample // get_world_size()
                       
    def __iter__(self):
        process_rank = get_rank()
        world_size = get_world_size()
        for url in self.urls:  
            # feed dummy pipe `wds.split_by_worker`, which is the default of workersplit, thus it'll not affect the results
            dataset = wds.WebDataset(url, nodesplitter=wds.split_by_worker).shuffle(1000).decode('pilrgb', handler=wds.warn_and_continue).to_tuple("jpg", "txt", handler=wds.warn_and_continue).map(self.trfs)
            for batch_id, sample in enumerate(dataset):
                # assign a independent batch for the gpu wrt. gpu_id (nodesplitter in here)
                if batch_id % world_size == process_rank: 
                    yield sample
                # skip the batch it doesn't belong to the gpu
                else: 
                    continue
```

3. The last step, take torch IterableDataset as a member in pl.DataModule
```
class Liaon115M(pl.LightningDataModule):

    def __init__(self, 
        data_dir: str = "path/to/dir", 
        split_ratio: list = [0.7, 0.1, 0.2], 
        img_transforms = None, 
        txt_transforms = None, 
        num_workers: int = 4, 
        batch_size: int = 16,
        num_epoch: int = 1,
        pin_memory: bool = False
    ):
        super().__init__()
        self.data_dir = Path(data_dir)
        self.split_ratio = split_ratio
        self.img_transforms = img_transforms
        self.txt_transforms = txt_transforms

        self.trfs = lambda tup : (self.img_transforms(tup[0]), self.txt_transforms(tup[1]))

        self.batch_size = batch_size
        self.num_epoch = num_epoch
        self.num_workers = num_workers
        self.pin_memory = pin_memory

        # laion 115M, success downloaded 104881705 records.. 
        
    # chking dataset is downloaded..
    def prepare_data(self): 
        with self.data_dir.open('r') as js_ptr:
            self.tar_dict = json.load(js_ptr)
        
        tar_lst = list(self.tar_dict.keys())
        n_shard = len(tar_lst)
        
        tra_ratio, val_ratio, _ = self.split_ratio
        self.tra_lst = tra_lst = tar_lst[ : int(n_shard * tra_ratio) ]
        self.val_lst = val_lst = tar_lst[ len(tra_lst) : len(tra_lst) + int(n_shard * val_ratio) ]
        self.tst_lst = tar_lst[ len(tra_lst) + len(val_lst) :  ]

    def _get_sample_num(self, tar_lst):
            cnt = 0
            for tar_key in tar_lst:
                cnt += self.tar_dict[tar_key]
            return cnt

    def setup(self, stage: str = 'train', rank=None, world_size=1):
        self.prepare_data()
        if stage == 'train':
            n_tra_sample = self._get_sample_num(self.tra_lst)
            print(f'Total sample in train ds : {n_tra_sample}')
            self.laion_train = Iter_ds(self.tra_lst, trfs=self.trfs, n_sample=n_tra_sample)
            n_val_sample = self._get_sample_num(self.val_lst)
            print(f'Total sample in valid ds : {n_val_sample}')
            self.laion_valid = Iter_ds(self.val_lst, trfs=self.trfs, n_sample=n_val_sample)
        else:
            n_tst_sample = self._get_sample_num(self.tst_lst)
            print(f'Total sample in test ds : {n_tst_sample}')
            self.laion_test = Iter_ds(self.tst_lst, trfs=self.trfs, n_sample=n_tst_sample)

    def train_dataloader(self):
        loader = torch.utils.data.DataLoader(
            self.laion_train, 
            batch_size=self.batch_size, 
            shuffle=False, 
            pin_memory=True, 
            num_workers=self.num_workers,
            prefetch_factor=2,
            drop_last=True
        )
        return loader 

    def val_dataloader(self):
        loader = torch.utils.data.DataLoader(
            self.laion_valid, 
            batch_size=self.batch_size, 
            shuffle=False, 
            pin_memory=True, 
            num_workers=self.num_workers,
            prefetch_factor=2,
            drop_last=True
        )
        return loader 
```

4. Since we setup the __len__ in iterableDataset, it will show the n_step and epoch in progress-bar

---

version of package 🥇 
1. webdataset : 0.2.86
2. pytorch-lightning : 2.2.1
3. torch : 2.2.0+cu118 

ok, now it's peace, no argument for torchlightning + webdataset, hopefully...

**Comment by HuangChiEn** [comment]
HAPPY CODING WITH TORCHLIGHTNING!!

**Comment by RobertLuo1** [comment]
Hi, Thanks for your code. Have you encountered that with the increased number of GPUs, the speed will get significantly slow?  I follow your template for training CC12M dataset and when I use 8 GPUs, the speed is 2.02iter/s. However, when I use 32 GPUs,  i.e., 4 machines, the speed is only 0.22iter/s.

**Comment by HuangChiEn** [comment]
> Hi, Thanks for your code. Have you encountered that with the increased number of GPUs, the speed will get significantly slow? I follow your template for training CC12M dataset and when I use 8 GPUs, the speed is 2.02iter/s. However, when I use 32 GPUs, i.e., 4 machines, the speed is only 0.22iter/s.

yes, it's possible in **multi-node** (multiple machine, each have multiple GPUs) training actually ~
it's a common issue we may encountered in **multi-node** environment ~

The first sad truth reveal in the following story : 
"Suppose i have 8 gpu in one node 0.6-iter/s, now i have 2 node (16 gpus), could i expect the 2xspeed (1.2-iter/s) ?
**In theory, yes**. However, it have several factors largely decrease the training speed : 
1. Network bandwidth between each node. If the sending data is in large size (see multi-node training, it'll sync the gradient for each network weights, ...etc.), it'll occupy the network bandwidth and largely decreasing the speed.
2. Each gpu is stay on a process, it may trigger 8 process per node (suppose one node have 8 gpus). Then the communication overhead between the processes will also decrease the speed.
so, it's completely possible slower then the previous one node setup.
"
I have the experience applied 2 node training on azure, since i setup the outter network routing (didn't go through by the virtual LAN) it even slower then one node.

If you're hosting 4 machines, it's lucky you can do something for the above factors i mentioned. If you can not do that, then my last suggestion is that you can use 4 machine to run the different experimental setup if you have.

Good luck, happy coding ~

**Comment by yifanlu0227** [comment]
Hi! If I using the `.with_epoch()` function for mulit-node training, should I also divide the `get_world_size()` inside `.with_epoch()`?

**Comment by HuangChiEn** [comment]
> Hi! If I using the `.with_epoch()` function for mulit-node training, should I also divide the `get_world_size()` inside `.with_epoch()`?
> 
> The upper-bound iteration in the progress bar never changed, no matter how many GPU I set up in the pl ddp trainer.

Unfortunately, i had tried several methods in document (`with_epoch` is included), but still cannot get the correct results even in single node (so i had gave up this way and create this issue with my solution). 
So, i can't extend those method in multi-node environment and provide the correct information. For this issue, you may ask the main contributor @tmbdev for help (he knows more in this case).

**Comment by RobertLuo1** [comment]
@HuangChiEn Thanks for your reply! When I set resampled=True, the speed is relatively normal. I think that parameter will benefit. However, I have another question is it possible to do shardshuffle=True? Since the code reveals that each shard can be viewed as a webdataset, it seems that the operation shuffling between shards is hard.

**Comment by HuangChiEn** [comment]
> resampled=True

sorry to reply late ~
i think `resample=True` can speedup by fetching the batched data **from their dataloader on process**. Namely, it *will not* fetch **the identical dataloader**, lead the progressbar show the longer step number.

For example.
Suppose we have 32 samples in dataloader and setup the batch size 4. If i have single gpu, it take `8=(32/4)` steps.
However, if i run on two gpu with the same batch size, it will still run the **8 steps** (because each gpu fetch from different dataloader)

// just my experience for sharing..

First, it depend on the property of dataset, not all dataset need to be shuffled. 
(1) The dataset need to shuffle : (i.e. sorted classification dataset)
    `dataset -> { [0~100, dog], [101~200, cat], ... so on }`, in this case we if doesn't shuffle the dataset, model may be overfit to the pattern that first learn all dog features, then learn all cat features, and so on.

(2) The dataset may not need to shuffle : (i.e. image-text pair dataset)
     Laion 115M crawl from website, it doesn't have such pattern like above. Each image-text pair is independent from its order in dataset.

---
In my case (Laion 115M), i don't strictly need to shuffle. But i first wrap webdataset as a single stream iterator, then the default sampler for torch multi-gpu `DistributedSampler`, it automatically shuffle **all data samples indices**, then take the subset for each gpu **(it equivalent to shard shuffle)**. 

The way you mentioned is also available ~ (shuffle urls or shuffle indices)

**Comment by RobertLuo1** [comment]
@HuangChiEn  Thanks for your reply! I think in your code, we can not set resample=True, since the for loop of the dataset. 
```
for batch_id, sample in enumerate(dataset):
```
However, without setting resampled=True, each time the work will try to read data from the disk, making it relatively slow. I wonder if there exists some efficient way to solve the problem by setting resampled=True and automatically breaking the for loop simultaneously.

**Comment by HuangChiEn** [comment]
> @HuangChiEn Thanks for your reply! I think in your code, we can not set resample=True, since the for loop of the dataset.
> 
> ```
> for batch_id, sample in enumerate(dataset):
> ```
> 
> However, without setting resampled=True, each time the work will try to read data from the disk, making it relatively slow. I wonder if there exists some efficient ways to solve the problem by setting resampled=True and automatically breaking the for loop simultaneously.

Umm.. let's say the for-loop simulate the general step in training loop. In torchlightning, `train_step` will fetch the batch data from for-loop underhood. So, i'm don't think `for batch_id, sample in enumerate(dataset)` become the issue. 

On the other hand, data sample is placed in disk basically. So, torch provide some flag, such as `pin_memory`, `prefetch_factor` to allow us load the samples in memory before feed it into GPU. 
> see all flags in https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader
If the slow speed cause by reading data from the disk, you can try to setup those flags. 

Lastly, if you want to setup `resampled=True` in this scenario, you may open a new issue to text the author of webdataset.

**Comment by sysuyy** [comment]
@HuangChiEn Thank you for your code！！！ But i get the following error, after the first step, do you have any idea?
![image](https://github.com/user-attachments/assets/cae36e83-7abf-47b4-a4e1-d84cf74f1124)

**Comment by HuangChiEn** [comment]
Could you list your package version ?
`pip list` (wanna chk `webdataset`, `torch`, `torchlightning`)
As the error message, it raised from webdataset :
![image](https://github.com/user-attachments/assets/6a30f344-526c-40ef-af38-62a4188d0416)

Is it possible to have a minimum reproduction :
```
# single worker
dataset = wds.WebDataset(url).shuffle(1000).decode('pilrgb', handler=wds.warn_and_continue).to_tuple("jpg", "txt", handler=wds.warn_and_continue).map(self.trfs)
for batch_id, sample in enumerate(dataset):
    pass
```
If it pass, then i guess it's caused by multiple workers.

**Comment by sysuyy** [comment]
thank you for reply !!!! @HuangChiEn 
torch                     2.4.1+cu124
webdataset                0.2.100
lightning                 2.4.0

when i change my code to
<img width="593" alt="image" src="https://github.com/user-attachments/assets/960e50c0-ae2e-4dab-bde1-8acd3d9e35d5">

it pass,
![Uploading image.png…]()

**Comment by HuangChiEn** [comment]
> thank you for reply !!!! @HuangChiEn torch 2.4.1+cu124 webdataset 0.2.100 lightning 2.4.0
> 
> when i change my code to <img alt="image" width="593" src="https://private-user-images.githubusercontent.com/68114531/370150877-960e50c0-ae2e-4dab-bde1-8acd3d9e35d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MTA2NDAsIm5iZiI6MTcyNzQxMDM0MCwicGF0aCI6Ii82ODExNDUzMS8zNzAxNTA4NzctOTYwZTUwYzAtYWUyZS00ZGFiLWJkZTEtOGFjZDNkOWUzNWQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDA0MTIyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxMTBkOTI1MTNkYmYzYzE1N2ZhNDczOWVkN2NhMjMzYjk1ZDc2MTNkYjI1NzkzNDlkOTI3NTgzZjljOGY3YTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UtQsizuGpaMCBTvUblBhTsgN7UUO-0MTwfJmT5IhsDk">
> 
> it pass,
> 
> and when i try to Alignment the version of webdatset to 0.2.86 the bug disappear, but i dont no why the next url is printed <img alt="image" width="1445" src="https://private-user-images.githubusercontent.com/68114531/370157693-ab3929af-9c28-47a3-ac14-5b72b24dd0bb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MTA2NDAsIm5iZiI6MTcyNzQxMDM0MCwicGF0aCI6Ii82ODExNDUzMS8zNzAxNTc2OTMtYWIzOTI5YWYtOWMyOC00N2EzLWFjMTQtNWI3MmIyNGRkMGJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDA0MTIyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhN2NjMzdjNzEwYzI2ZjliNTE4YjJhMzQ2MTM4NmMyZTg1NTk1NjQ3YmQ4NTY3NzE0MTVjNmRkZTBhZmMxNGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BRDfA_mYwTKcodXJhne-EcPMBdQRVNo3kQWT6ybcIYA">

Hello, glad to hear that the bug disappear. You can comment out the `print(urlll...)` to filter out the message. While this `print` function place in there without the rank condition (`if rank==0: print(...)`), so all of the rank will print this information. 

As your screen shot, each process actually will first received each url, but the code snippet below will filter out the url not belong to the rank : 
```
def __iter__(self):
        process_rank = get_rank()
        world_size = get_world_size()
        for url in self.urls:  
            # all url will be received by each process!!
            dataset = wds.WebDataset(url, nodesplitter=wds.split_by_worker).shuffle(1000).decode('pilrgb', handler=wds.warn_and_continue).to_tuple("jpg", "txt", handler=wds.warn_and_continue).map(self.trfs)
            for batch_id, sample in enumerate(dataset):
                if batch_id % world_size == process_rank: 
                    yield sample

                # drop out the batch sample which not belong to this process, 
                # respect with the process rank!!
                else: 
                    continue
```
